nexus/client.go

```diff
@@ -0,0 +1,629 @@
+}
+
+// OperationCompletionSuccessful is input for [Client.DeliverCompletion], used to deliver successful operation results.
+type OperationCompletionSuccessful struct {
+	// Header to send in the completion request.
+	Header http.Header
+	// Body to send in the completion HTTP request.
+	// If it implements `io.Closer` it will automatically be closed by the client.
+	Body io.Reader
+}
+
+// NewSuccessfulOperationCompletion constructs an [OperationCompletionSuccessful] from a JSONable value.
+// Marshals the provided value to JSON using [json.Marshal] and sets the proper Content-Type header.
+func NewSuccessfulOperationCompletion(v any) (*OperationCompletionSuccessful, error) {
+	b, err := json.Marshal(v)
+	if err != nil {
+		return nil, err
+	}
+
+	header := make(http.Header, 1)
+	header.Set(headerContentType, contentTypeJSON)
+
+	return &OperationCompletionSuccessful{
+		Header: header,
+		Body:   io.NopCloser(bytes.NewReader(b)),
+	}, nil
+}
+
+// OperationCompletionUnsuccessful is input for [Client.DeliverCompletion], used to deliver unsuccessful operation
+// results.
+type OperationCompletionUnsuccessful struct {
+	// Header to send in the completion request.
+	Header http.Header
+	// State of the operation, should be failed or canceled.
+	State OperationState
+	// Failure object to send with the completion.
+	Failure *Failure
+}
+
+func (c *OperationCompletionSuccessful) applyToHTTPRequest(request *http.Request, client *Client) error {
+	if c.Header != nil {
+		request.Header = c.Header.Clone()
+	}
+	request.Header.Set(headerOperationState, string(OperationStateSucceeded))
+	if closer, ok := c.Body.(io.ReadCloser); ok {
+		request.Body = closer
+	} else {
+		request.Body = io.NopCloser(c.Body)
+	}
+	return nil
+}
+
+// Close implements the io.Closer interface.
+func (c *OperationCompletionSuccessful) Close() error {
+	if closer, ok := c.Body.(io.Closer); ok {
+		return closer.Close()
+	}
+	return nil
+}
+
+func (c *OperationCompletionUnsuccessful) applyToHTTPRequest(request *http.Request, client *Client) error {
+	if c.Header != nil {
+		request.Header = c.Header.Clone()
+	}
+	request.Header.Set(headerOperationState, string(c.State))
+	request.Header.Set(headerContentType, contentTypeJSON)
+
+	b, err := client.Options.Marshaler(c.Failure)
+	if err != nil {
+		return err
+	}
+
+	request.Body = io.NopCloser(bytes.NewReader(b))
+	return nil
+}
+
+// Close implements the io.Closer interface.
+func (c *OperationCompletionUnsuccessful) Close() error {
+	return nil
+}
+
+// readAndReplaceBody reads the response body in its entirety and closes it, and then replaces the original response
+// body with an in-memory buffer.
+func readAndReplaceBody(response *http.Response) ([]byte, error) {
+	responseBody := response.Body
+	defer responseBody.Close()
+	body, err := io.ReadAll(responseBody)
+	if err != nil {
+		return nil, err
+	}
+	response.Body = io.NopCloser(bytes.NewReader(body))
+	return body, nil
```

Comment:
I know about partial reads but I don't need to replace the body on failure because I never return the response on error here.

Reply:
Also take into consideration that there's no way to indicate that reading the body failed, so a consumer won't know that there's partial body.

Reply:
> I never return the response on error here

The problem here is that you have chosen to break this simple code out into a method instead of inline it (which most Go code would for such a simple operation). Since you have chosen this, I think you've given up your right to know what the caller does with the object they passed. This is how bugs often happen, little unnecessary utilities making implicit assumptions about their caller. At the very least add docs on this method that say the response body is not usable after this method if an error is returned.